app/src/main/java/de/westnordost/streetcomplete/map/tangram/CameraManager.kt

```diff
@@ -141,7 +141,10 @@ class CameraManager(private val c: MapController, private val contentResolver: C
         animator.setValues(*propValues.toTypedArray())
         animator.duration = duration
         animator.interpolator = interpolator
-        animator.addListener(onEnd = this::unassignAnimator)
+        animator.addListener(onEnd = {
+            animate(it as ValueAnimator)
+            unassignAnimator(it)
+        })
```

Comment:
Actually, if you wonder why this is done seemingly so complicated with the "lastAnimator": This is because there should only be one call to `pushCameraPositionToController` (per animation frame), even if multiple animations of different properties are running at the same time.
Initial implementation was to just push the camera position to the controller in this listener, but this resulted in the animated values overwriting each other. So, the solution was that only one animator will push the position, and that one is the animator that will run the longest.

So, the additional call to `animate` on end of a single object animator should rather go to the "lastAnimator".

Reply:
Hm. So, I tried doing this in the latest.

1. Can you check that what I did is what you intended?
2. Unfortunately, this didn't work.

Reply:
Hm yes, this is what I intended.

So, if that means that every object animator will stop just before reaching the final value (sounds like a bug in the android animation system to me), then what ought to fix it is additionally to how the code is now, `onEnd` of *every* object animator, set the animated property to the final value, but do not push that to the camera controller because the "lastAnimator" will do that.

Reply:
I'm not sure which would be more work, but the way Red Moon handles this is to use a single animator for everything, and if there's an animation currently running when a new one starts (eg, filter is turned off during fade in animation), we just set the values to start there.

If you actually want to be running multiple animators at once, I think there's an AnimatorCoordinator or something like that.

Reply:
True, but these animations are be started at different points in time as well as have different interpolators and durations.

Reply:
The `lastAnimator` bit is easy enough for me to understand. The part I'm actually having trouble with is figuring out how updated CameraPosition values actually get pushed to `_tangramCamera` from the animators. Obviously it happens in `applyCameraUpdate`, but that only gets called in `updateCamera`, which can be called from several places, none of which seem to be `onAnimationUpdate`. However, understanding this is starting to take too much time, so _I'm going to black box it and just accept that the values are getting where they need to go._ I want to fix the "almost `0f`" bug, not re-architect your tangram wrappers :rofl:

Anyway, back to `pushCameraPositionToController`, the flow of which I do understand. Here's what I'm thinking about:

- Remove all `pushCameraPositionToController` calls from the current animators, leaving them responsible only for updating the camera position.
- Remove the `lastAnimator` concept.
- Add a new [TimeAnimator](https://developer.android.com/reference/android/animation/TimeAnimator) whose only purpose is calling `pushCameraPositionToController`
- Instead of checking `lastAnimatorEndTime < endTime`,  compare the `timeAnimatorEndTime`. If the new end time is later than the TimeAnimator's end time extend the TimeAnimator (ie, keep pushing the camera position, as long as any animations are currently running).

Reply:
> The part I'm actually having trouble with is figuring out how updated CameraPosition values actually get pushed to _tangramCamera from the animators

ObjectAnimators do this automatically. ObjectAnimators directly manipulate the properties of the animated object (the `target`), in this case _tangramCamera. `applyCameraUpdate` is just to apply an update once and push it to tangram right away.

Reply:
> Add a new TimeAnimator that exclusively calls pushCameraPositionToController
> Instead of checking lastAnimatorEndTime < endTime, compare the timeAnimatorEndTime. If the new end time is later than the TimeAnimator's end time, extend the TimeAnimator (ie, keep pushing the camera position, as long as any animations are currently running).

I didn't know about `TimeAnimator`, this sounds like the right thing to use. Using it will not really lead to less code, but probably a bit cleaner code.

> Remove all pushCameraPositionToController calls from the current animators, leaving them responsible only for updating the camera position.

The current animators do not call `pushCameraPositionToController`

Reply:
> ObjectAnimators directly manipulate the properties of the animated object (the target)

d'oh :man_facepalming:. I do vaguely remember reading this now, and wonder why I had Red Moon manually applying the animatedValue instead :man_shrugging: 

> The current animators do not call `pushCameraPositionToController`

They call `animate`, which calls `pushCameraPositionToController` :P

Reply:
> They call animate (or rather, the lastAnimator does), which calls pushCameraPositionToController :P

Only the "lastAnimator" does this. If you change the implementation so that a TimeAnimator will do this, then none of the ObjectAnimators will do this anymore.

Reply:
I think we're on the same page here, just talking past each other.

Reply:
ok

Reply:
Okay, here's what I'm confused about now: `adb logcat -s 'SC:*'` with the code I just pushed.

The correct values are pushed to `_tangramCamera`, but then some time *after* `onEnd`, incorrect values overwrite them. `onEnd` should be happening after the animation is finished, so where are the off-zero values coming from?

Reply:
Out of time this morning but I'm narrowing in on the culprit with 5f27b54.

I'm leaving the commit history intact for the moment so you can checkout and see for yourself (will squash later), but the diff between master and 5f27b54 is *only* adding the `onUpdateListener`.

As far as I can tell, the only difference between setting the values manually and automatically with the animator target is that I'm setting them conditionally. So, maybe running multiple animators with the same target is the cause of the bug?

Reply:
One thing I noticed is that the `_tangramCamera` as a target of the animators is one of the few times that accessing those values is not `synchronized(_tangramCamera)`. I tried adding that into the `onAnimationUpdate` call and it caused crashes, though.

Reply:
I lost track here a little. You add line 144-153 because you want to push the animator end value(s) to the `tangramCamera`. But it doesn't work - it is overwritten.
Let's not waste any more time with this and instead use the fix you did from the very beginning, perhaps with a little comment.

To be honest, l don't believe that the animator system in Android is so crooked that the animator will stop right before reaching the exact end value. But even if it did, if the workaround for this turns out to add even more complexity to it (and the workaround also doesn't work and needs to be work-arounded), then it is not worth it to fix.

From a UX perspective, the user would expect that the "to north" button will treat very small (=barely visible) deviations from North as "it is north already" anyway.

Reply:
> I lost track here a little.

Commits need squashing and logging removed, but the current implementation works.

> You add line 144-153 because you want to push the animator end value(s) to the `tangramCamera`

I was trying to understand what's going on. Does it reach the final value and then overshoot? Does it get overridden by a different animator? etc.

Originally I only had some logging. Then I started adding random stuff (well, educated guesses) to see how it would respond. This one happened to work. I still don't quite understand why.

> l don't believe that the animator system in Android is so crooked that the animator will stop right before reaching the exact end value.

It's possible I'm misremembering, and it's actually the time frame that it overshoots slightly (eg, animate for 320ms even though you said 300).

> From a UX perspective, the user would expect that the "to north" button will treat very small (=barely visible) deviations from North as "it is north already" anyway.

Seems reasonable, regardless of whether the bug is fixed. I was using `0.02f`. That caught ~90% of cases, but sometimes the "almost `0f` bug still hit. Should I bump it up to `0.03f`?

> it is not worth it to fix

The current CameraManager code seems to work (end values are correct). It's +8 lines that don't seem like they should be necessary. Should I keep it or leave it?

Reply:
> The current CameraManager code seems to work (end values are correct). It's +8 lines that don't seem like they should be necessary. Should I keep it or leave it?

Rather if you removed them again. It is a workaround that "happens to work" for a bug perhaps exists (in Android). Not working around this possible bug has no real downsides, so less code here is better.

> 0.03f

I'd say whatever feels natural as a user. 0.03f would be around 5°, might be a bit too much.

Reply:
I went with `val margin = 0.025f`. While I am able to start at true north, rotate slightly, and then tap the compass and have it tilt instead of moving back to true north, I could not do it reliably (maybe once every 5 or 10 tries), and this was with me *trying* to do that, so in actual operation I think it should happen basically never. Also pulling it out into a val makes it really easy to change in the future.